### PR TITLE
Kevinl/forget password 

### PR DESCRIPTION
**Pull request by KevinLi3**
Forgot password functionality works 
Screen Shot 2023-08-14 at 4 48 12 PM
Screen Shot 2023-08-14 at 4 53 04 PM
Screen Shot 2023-08-14 at 4 53 23 PM

Screen Shot 2023-08-14 at 4 56 23 PM
3-08-14 at 4 55 48 PM" src="https://github.com/nehageorge/xchange/assets/50978011/0ccbc6f0-ea53-47f7-8d90-455617cc6947">

### CODE REVIEW

File: api/app.py [hunk 1]
```diff
@@ -1,24 +1,35 @@
...
+
+
 db = SQLAlchemy(app)
 migrate = Migrate(app, db)
-
+app.config['MAIL_SERVER'] = 'smtp-mail.outlook.com'
+app.config['MAIL_PORT'] = 587
+app.config['MAIL_USE_TLS'] = True
+app.config['MAIL_USERNAME'] = 'xchangeuw@outlook.com'
+app.config['MAIL_PASSWORD'] = '12345Abc@'
```
Comment (AndreaMiranda99) [inline api/app.py]: Is this a placeholder password ? 

For the real password ensure that it is an environment variable that we use here so we dont expose these values in git.

File: api/app.py [hunk 2]
```diff
@@ -53,6 +64,30 @@ class User(db.Model):
...
+    def get_token(self,expires_sec=600):
+        # serial = Serializer(app.config['SECRET_KEY'], expires_in=expires_sec)
+        # return serial.dumps({'user_id':self.id}).decode('utf-8')
+        return jwt.encode({'reset_password': self.id,
+                           'exp':    time.time() + expires_sec},
+                           key=os.getenv('SECRET_KEY_FLASK'), algorithm="HS256")
+
+    @staticmethod
+    def verify_token(token):
+        # serial = Serializer(app.config['SECRET_KEY'])
```
Comment (AndreaMiranda99) [inline api/app.py]: Commented code should be removed if it is no longer in use.

File: api/app.py [hunk 3]
```diff
@@ -53,6 +64,30 @@ class User(db.Model):
     password = db.Column(db.String(128))
     is_admin = db.Column(db.Integer)
 
+    def get_token(self,expires_sec=600):
+        # serial = Serializer(app.config['SECRET_KEY'], expires_in=expires_sec)
+        # return serial.dumps({'user_id':self.id}).decode('utf-8')
+        return jwt.encode({'reset_password': self.id,
```
Comment (AndreaMiranda99) [inline api/app.py:71]: What if this throws an error or returns None ? The email that is sent wouldn't have a token in it then.

File: api/app.py [hunk 4]
```diff
@@ -166,6 +201,62 @@ def signup():
...
+def forgot_password():
+    if request.method == 'POST':
+        email = request.form['email']
+
+        result = User.query.filter(User.email == email).first()
+        if result is None:
+            e = "This email is not registered."
+            return redirect(url_for('login_error', problem=str(e)))
+        send_mail(result)
+        #TODO: make this actually show up
```
Comment (AndreaMiranda99) [inline api/app.py]: Does it show up ?

File: api/app.py [hunk 5]
```diff
@@ -166,6 +201,62 @@ def signup():
...
+    
+    {url_for('forgot_password_success', token=token,_external=True).replace("http://127.0.0.1:5000", "http://127.0.0.1:3000")}
+    If you didn't send a password reset request, please ignore this message.
+
+    '''
+
+    mail.send(msg)
+
+@app.route('/forgot_password', methods=['GET','POST'])
+#Need to use UserBuilder, or do the same salting protocol as in UserBuilder
```
Comment (AndreaMiranda99) [inline api/app.py]: There is not other mention of UserBuilder from the new code added in this PR, Is this no longer required ? I do not see it used in this route.

File: api/app.py [hunk 6]
```diff
@@ -166,6 +201,62 @@ def signup():
...
+        flash('Reset request sent. Check your mail.', 'success')
+        return redirect(url_for('login'))
+    else:
+        return jsonify("")
+
+#Need to pass in token
+@app.route('/forgot_password_success/<token>', methods=['GET', 'POST'])
+def forgot_password_success(token):
+    print("Token: " + str(token))
+    # if request.method == 'POST':
```
Comment (AndreaMiranda99) [inline api/app.py]: Remove commented code if no longer needed.

**Review by AndreaMiranda99 — CHANGES_REQUESTED** [review]

File: api/app.py (hunk 1 above)
Comment (KevinLi3) [reply api/app.py]: Resolved, put into Env variable

File: api/app.py (hunk 2 above)
Comment (KevinLi3) [reply api/app.py]: Resolved

File: api/app.py (hunk 3 above)
Comment (KevinLi3) [reply api/app.py:71]: Resolved

File: api/app.py (hunk 4 above)
Comment (KevinLi3) [reply api/app.py]: Yes

File: api/app.py (hunk 5 above)
Comment (KevinLi3) [reply api/app.py]: No longer needed

File: api/app.py (hunk 6 above)
Comment (KevinLi3) [reply api/app.py]: Resolved

**Review by nehageorge — APPROVED** [review]

**Review by AndreaMiranda99 — APPROVED** [review]

**Pull request merged**